stan-dev/math
# Feature/issue exponential covariance function rev/prim

#### Submission Checklist


#### Summary:
This adds exponential covariance function in rev, and prim, with ARD support in prim. It will be easy to switch around between kernels, per the suggestion on discourse.

I only added `rev` support for one input x vector, and combinations of var/double for sigma and length_scale. If I run some models and it's slow, or users report it being slow, I'll come back after testing and `rev` some more stuff.

#### How to Verify:
I encourage you to check my arithmetic and derivatives, but here are the test cases for prim, and then rev, where v and d are `stan::math::var` and `std::double`, respectively:

rev:
all combinations of v,d (with 3 possible cells) the same test cases for vector inputs,
nan/error training, and vars on the stack.

prim (copy and pasted from previous PR):
vec double
vec double double
vec double ARD
vec eigen
vec eigen eigen
vec eigen ARD
vec double double ARD
rvec eigen
rvec eigen ARD
vec eigen rvec eigen
vec eigen rvec eigen ARD
rvec eigen rvec eigen
rvec eigen rvec eigen ARD
eigen mixed
eigen mixed ARD
double domain err training
nan error training
differing x lengths

#### Documentation:
doxygen and GMPL
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
Copywrite<2018><Andre Zapico>

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
